### PR TITLE
distribution: errors: do not access the errors slice if it's empty

### PR DESCRIPTION
**Pull request by runcom**
Fix https://github.com/docker/docker/issues/21027
- cherry-pick from 1.10.3 branch: 0186f4d4223a094a050d06f456355da3ae431468
- add token service test suite
- add integration test (missing in 1.10.3 branch)

@aaronlehmann @tiborvass @thaJeztah 

Signed-off-by: Antonio Murdaca runcom@redhat.com

### CODE REVIEW

**Comment by thaJeztah** [comment]
Thanks @runcom

**Comment by runcom** [comment]
:heart: :whale:

**Comment by thaJeztah** [comment]
Looks like you're missing some changes;

```
10:07:22 ---> Making bundle: binary (in bundles/1.11.0-dev/binary)
10:07:23 Building: bundles/1.11.0-dev/binary/docker-1.11.0-dev.exe
10:07:37 # github.com/docker/docker/cliconfig/credentials
10:07:37 cliconfig\credentials\native_store.go:82: auth.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:37 cliconfig\credentials\native_store.go:82: creds.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:37 cliconfig\credentials\native_store.go:96: ac.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:37 cliconfig\credentials\native_store.go:96: creds.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:37 cliconfig\credentials\native_store.go:110: authConfig.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:37 cliconfig\credentials\native_store.go:125: config.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:37 cliconfig\credentials\native_store.go:127: config.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:37 cliconfig\credentials\native_store.go:173: ret.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:39 # github.com/docker/docker/runconfig/opts
10:07:39 runconfig\opts\parse.go:413: unknown container.HostConfig field 'UsernsMode' in struct literal
10:07:39 # github.com/docker/docker/registry
10:07:39 registry\auth.go:87: lcs.authConfig.IdentityToken undefined (type *types.AuthConfig has no field or method IdentityToken)
10:07:39 registry\auth.go:91: lcs.authConfig.IdentityToken undefined (type *types.AuthConfig has no field or method IdentityToken)
10:07:39 registry\auth.go:167: credentialAuthConfig.IdentityToken undefined (type types.AuthConfig has no field or method IdentityToken)
10:07:46 # github.com/docker/docker/api/server/router/system
10:07:46 api\server\router\system\system_routes.go:124: unknown types.AuthResponse field 'IdentityToken' in struct literal
10:08:50 + ec=1
10:08:50 + set +x
10:08:50 ERROR: Build of binary on Windows failed
```

**Comment by runcom** [comment]
weird, rebased and push forced

**Comment by runcom** [comment]
mmm win2lin failures seems not related - my patch is correctly rebased, something else is going on :confused:

**Comment by runcom** [comment]
I'm actually thinking that the 1.10.3 bump branch should be merged first in master - then I'll add tests, @tiborvass ?

**Comment by runcom** [comment]
This should be good to review now

File: distribution/errors.go [hunk 1]
```diff
@@ -84,7 +84,9 @@ func continueOnError(err error) bool {
 func retryOnError(err error) error {
 	switch v := err.(type) {
 	case errcode.Errors:
-		return retryOnError(v[0])
+		if len(v) != 0 {
+			return retryOnError(v[0])
+		}
```
Comment (runcom) [inline distribution/errors.go]: this check won't be necessary after https://github.com/docker/distribution/pull/1531 and distribution vendor is updated /cc @aaronlehmann

File: integration-cli/docker_cli_push_test.go [hunk 2]
```diff
@@ -546,3 +548,33 @@ func (s *DockerSuite) TestPushToCentralRegistryUnauthorized(c *check.C) {
...
+		w.Header().Set("Content-Type", "application/json")
+		// this will make the daemon panics if no check is performed in retryOnError
+		w.Write([]byte(`{"error": "unauthorized"}`))
+	}))
+	defer ts.Close()
+	s.setupRegistryWithTokenService(c, ts.URL)
+	repoName := fmt.Sprintf("%s/busybox", privateRegistryURL)
+	dockerCmd(c, "tag", "busybox", repoName)
+	out, _, err := dockerCmdWithError("push", repoName)
+	c.Assert(err, check.NotNil, check.Commentf(out))
```
Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: this should also check that I get unexpected http error in light of https://github.com/docker/distribution/pull/1531 /cc @aaronlehmann

File: distribution/errors.go (hunk 1 above)
Comment (aaronlehmann) [inline distribution/errors.go]: > this check won't be necessary after docker/distribution#1531 and distribution vendor is updated

Yep, was going to point this out. But it's defensive to have the check anyway, so I like this as-is.

Comment (runcom) [inline distribution/errors.go]: right, we would probably remember that we should change the test in https://github.com/docker/docker/pull/21212#discussion_r56320243

**Comment by aaronlehmann** [comment]
LGTM

This must be included in 1.11.0 to bring it to parity with 1.10.3, so let's merge this as-is, and when we vendor an updated `docker/distribution` we can update the tests to check for the new error value.

File: integration-cli/docker_cli_push_test.go (hunk 2 above)
Comment (runcom) [inline integration-cli/docker_cli_push_test.go]: I'll make a follow up PR just to check the output string so when we'll vendor docker/distribution, this test will fail and we can instantly fix it w/o forgetting :)

**Comment by tiborvass** [comment]
LGTM thanks for catching this @runcom

**Comment by calavera** [comment]
LGTM

**Pull request merged**